File: android-kotlin/QuickStartTasks/app/src/main/java/live/ditto/quickstart/tasks/list/TasksListScreenViewModel.kt
package live.ditto.quickstart.tasks.list

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import live.ditto.DittoError
import live.ditto.quickstart.tasks.DittoManager
import live.ditto.quickstart.tasks.TasksApplication
import live.ditto.quickstart.tasks.data.Task

// Removed DataStore persistence - transport states now default to true on app start

class TasksListScreenViewModel : ViewModel() {

    companion object {
        private const val TAG = "TasksListScreenViewModel"
        private const val QUERY = "SELECT * FROM tasks WHERE NOT deleted ORDER BY title ASC LIMIT 50"
        private const val COUNT_QUERY = "SELECT COUNT(*) as result FROM tasks WHERE NOT deleted"
    }

    private val dittoManager: DittoManager = TasksApplication.getDittoManager()

    val tasks: MutableLiveData<List<Task>> = MutableLiveData(emptyList())

    private val _syncEnabled = MutableLiveData(true)
    val syncEnabled: LiveData<Boolean> = _syncEnabled

    private val _bluetoothEnabled = MutableLiveData(true)
    val bluetoothEnabled: LiveData<Boolean> = _bluetoothEnabled

    private val _lanEnabled = MutableLiveData(true)
    val lanEnabled: LiveData<Boolean> = _lanEnabled

    private val _wifiAwareEnabled = MutableLiveData(true)
    val wifiAwareEnabled: LiveData<Boolean> = _wifiAwareEnabled

    val count: MutableLiveData<Int> = MutableLiveData(0)

    fun setSyncEnabled(enabled: Boolean) {
        viewModelScope.launch {
            _syncEnabled.value = enabled

            if (enabled && !dittoManager.isSyncActive()) {
                try {
                    dittoManager.startSync()
                } catch (e: DittoError) {
                    Log.e(TAG, "Unable to start sync", e)
                }
            } else if (!enabled && dittoManager.isSyncActive()) {
                try {
                    dittoManager.stopSync()
                } catch (e: DittoError) {
                    Log.e(TAG, "Unable to stop sync", e)
                }
            }
        }
    }

    fun toggleBluetoothLE() {
        val newValue = !(_bluetoothEnabled.value ?: true)
        _bluetoothEnabled.value = newValue
        dittoManager.toggleBluetoothLE()
    }

    fun toggleLAN() {
        val newValue = !(_lanEnabled.value ?: true)
        _lanEnabled.value = newValue
        dittoManager.toggleLAN()
    }

    fun toggleWifiAware() {
        val newValue = !(_wifiAwareEnabled.value ?: true)
        _wifiAwareEnabled.value = newValue
        dittoManager.toggleWifiAware()
    }

    init {
        viewModelScope.launch {
            // Wait for Ditto to be initialized before setting up observers
            while (!dittoManager.isDittoInitialized()) {
                kotlinx.coroutines.delay(100) // Wait 100ms before checking again
            }

            populateTasksCollection()
            setupObservers()

            // Start sync by default (all transports default to enabled)
            setSyncEnabled(true)
        }
    }

    private fun setupObservers() {
        viewModelScope.launch {
            // Observe tasks using Flow
            dittoManager.liveQueryAsFlow(QUERY, emptyMap()).collect { result ->
                val list = result.items.map { item -> Task.fromJson(item.jsonString()) }
                tasks.postValue(list)
            }
        }

        viewModelScope.launch {
            // Observe count using Flow
            dittoManager.liveQueryAsFlow(COUNT_QUERY, emptyMap()).collect { result ->
                result.items.forEach {
                    count.postValue(it.value["result"] as Int)
                }
            }
        }
    }

    // Add initial tasks to the collection if they have not already been added.
    private fun populateTasksCollection() {
        viewModelScope.launch {
            if (!dittoManager.isDittoInitialized()) {
                Log.w(TAG, "Cannot populate tasks - Ditto not initialized")
                return@launch
            }

            val tasks = listOf(
                Task("50191411-4C46-4940-8B72-5F8017A04FA7", "Buy groceries"),
                Task("6DA283DA-8CFE-4526-A6FA-D385089364E5", "Clean the kitchen"),
                Task("5303DDF8-0E72-4FEB-9E82-4B007E5797F0", "Schedule dentist appointment"),
                Task("38411F1B-6B49-4346-90C3-0B16CE97E174", "Pay bills")
            )

            tasks.forEach { task ->
                dittoManager.executeQuery(
                    "INSERT INTO tasks INITIAL DOCUMENTS (:task)",
                    mapOf(
                        "task" to mapOf(
                            "_id" to task._id,
                            "title" to task.title,
                            "done" to task.done,
                            "deleted" to task.deleted,
                        )
                    )
                ).onFailure { e ->
                    Log.e(TAG, "Unable to insert initial document", e)
                }
            }
        }
    }

    fun toggle(taskId: String) {
        viewModelScope.launch {
            if (!dittoManager.isDittoInitialized()) {
                Log.w(TAG, "Cannot toggle task - Ditto not initialized")
                return@launch
            }

            try {
                val docResult = dittoManager.executeQuery(
                    "SELECT * FROM tasks WHERE _id = :_id AND NOT deleted",
                    mapOf("_id" to taskId)
                )

                docResult.onSuccess { result ->
                    val doc = result.items.first()
                    val done = doc.value["done"] as Boolean

                    dittoManager.executeQuery(
                        "UPDATE tasks SET done = :toggled WHERE _id = :_id AND NOT deleted",
                        mapOf(
                            "toggled" to !done,
                            "_id" to taskId
                        )
                    ).onFailure { e ->
                        Log.e(TAG, "Unable to toggle done state", e)
                    }
                }.onFailure { e ->
                    Log.e(TAG, "Unable to find task", e)
                }
            } catch (e: Exception) {
                Log.e(TAG, "Unable to toggle done state", e)
            }
        }
    }

    fun delete(taskId: String) {
        viewModelScope.launch {
            if (!dittoManager.isDittoInitialized()) {
                Log.w(TAG, "Cannot delete task - Ditto not initialized")
                return@launch
            }

            dittoManager.executeQuery(
                "UPDATE tasks SET deleted = true WHERE _id = :id",
                mapOf("id" to taskId)
            ).onFailure { e ->
                Log.e(TAG, "Unable to set deleted=true", e)
            }
        }
    }

    fun deleteAllIncomplete() {
        viewModelScope.launch {
            if (!dittoManager.isDittoInitialized()) {
                Log.w(TAG, "Cannot delete tasks - Ditto not initialized")
                return@launch
            }

            // Use DQL DELETE to permanently delete all incomplete tasks
            dittoManager.executeQuery(
                "DELETE FROM tasks WHERE done = false",
                emptyMap()
            ).onSuccess { result ->
                Log.d(TAG, "Successfully deleted all incomplete tasks")
            }.onFailure { e ->
                Log.e(TAG, "Unable to delete incomplete tasks", e)
            }
        }
    }

    fun evictAllDeleted() {
        viewModelScope.launch {
            if (!dittoManager.isDittoInitialized()) {
                Log.w(TAG, "Cannot evict tasks - Ditto not initialized")
                return@launch
            }

            dittoManager.executeQuery(
                "EVICT FROM tasks WHERE deleted = true",
                emptyMap()
            ).onSuccess {
                Log.d(TAG, "Successfully evicted all deleted tasks")
            }.onFailure { e ->
                Log.e(TAG, "Unable to evict deleted tasks", e)
            }
        }
    }
}

File: swift/Tasks/ViewModels/TasksListScreenViewModel.swift
import Combine
import DittoSwift
import SwiftUI

/// View model for TasksListScreen
@MainActor
class TasksListScreenViewModel: ObservableObject {
    @Published var tasks = [TaskModel]()
    @Published var count: Int = 0
    @Published var isPresentingEditScreen: Bool = false
    @Published var isPresentingBulkAddScreen: Bool = false
    @Published var isPresentingConfigurationView: Bool = false
    private(set) var taskToEdit: TaskModel?

    private let ditto = DittoManager.shared.ditto
    private var subscription: DittoSyncSubscription?
    private var taskObserver: DittoStoreObserver?
    private var countObserver: DittoStoreObserver?


    init() {
        populateTasksCollection()

        startTaskObserver()
        startCountObserver()
    }

    deinit {
        subscription?.cancel()
        subscription = nil

        taskObserver?.cancel()
        taskObserver = nil

        countObserver?.cancel()
        countObserver = nil

        if ditto.sync.isActive {
            ditto.sync.stop()
        }
    }
    
    func toggleBluetoothLE() {
        ditto.transportConfig.peerToPeer.bluetoothLE.isEnabled.toggle()
    }
    
    func toggleLAN() {
        ditto.transportConfig.peerToPeer.lan.isEnabled.toggle()
    }
    
    func toggleAWDL() {
        ditto.transportConfig.peerToPeer.awdl.isEnabled.toggle()
    }

    func setSyncEnabled(_ newValue: Bool) throws {
        if !ditto.sync.isActive && newValue {
            try startSync()
        } else if ditto.sync.isActive && !newValue {
            stopSync()
        }
    }

    private func startSync() throws {
        let query = "SELECT * from tasks"
        
        try ditto.sync.start()
        subscription = try ditto.sync.registerSubscription(query: query)
    }

    private func stopSync() {
        subscription?.cancel()
        subscription = nil

        ditto.sync.stop()
    }
    
    private func startTaskObserver() {
        let query = "SELECT * FROM tasks WHERE NOT deleted ORDER BY title ASC LIMIT 50"
        
        taskObserver = try? ditto.store.registerObserver(query: query) {
            [weak self] result in
            guard let self = self else { return }
            self.tasks = result.items.compactMap {
                TaskModel($0.jsonData())
            }
        }
    }
    
    private func startCountObserver() {
        let query = "SELECT COUNT(*) as result FROM tasks where NOT deleted"
        
        countObserver = try? ditto.store.registerObserver(query: query) {
            [weak self] result in
            guard let self = self else { return }

            self.count = (result.items.first?.value as? [String: Any?])?["result"] as? Int ?? 0
        }
    }

    func toggleComplete(task: TaskModel) {
        Task {
            let done = !task.done
            let query = """
                UPDATE tasks
                SET done = :done
                WHERE _id == :_id
                """

            do {
                try await ditto.store.execute(
                    query: query,
                    arguments: ["done": done, "_id": task._id]
                )
            } catch {
                print(
                    "TaskListScreenVM.\(#function) - ERROR toggling task: \(error.localizedDescription)"
                )
            }
        }
    }

    nonisolated func saveEditedTask(_ task: TaskModel) {
        Task {
            let query = """
                UPDATE tasks SET
                    title = :title,
                    done = :done,
                    deleted = :deleted
                WHERE _id == :_id
                """

            do {
                try await ditto.store.execute(
                    query: query,
                    arguments: [
                        "title": task.title,
                        "done": task.done,
                        "deleted": task.deleted,
                        "_id": task._id
                    ]
                )
            } catch {
                print(
                    "TaskListScreenVM.\(#function) - ERROR updating task: \(error.localizedDescription)"
                )
            }
        }
    }

    nonisolated func saveNewTask(_ task: TaskModel) {
        Task {
            let newTask = task.value
            let query = "INSERT INTO tasks DOCUMENTS (:newTask)"

            do {
                try await ditto.store.execute(
                    query: query, arguments: ["newTask": newTask])
            } catch {
                print(
                    "TaskListScreenVM.\(#function) - ERROR creating new task: \(error.localizedDescription)"
                )
            }
        }
    }

    nonisolated func deleteTask(_ task: TaskModel) {
        Task {
            let query = "UPDATE tasks SET deleted = true WHERE _id = :_id"
            do {
                try await ditto.store.execute(
                    query: query, arguments: ["_id": task._id])
            } catch {
                print(
                    "TaskListScreenVM.\(#function) - ERROR deleting task: \(error.localizedDescription)"
                )
            }
        }
    }
    
    nonisolated func deleteAllTasks() {
        Task {
            let query = "UPDATE tasks SET deleted = true WHERE done = false"
            do {
                try await ditto.store.execute(query: query)
            } catch {
                print(
                    "TaskListScreenVM.\(#function) - ERROR deleting tasks: \(error.localizedDescription)"
                )
            }
        }
    }
    
    nonisolated func evictAllDeletedTasks() {
        Task {
            let query = "EVICT FROM tasks WHERE deleted = true"
            do {
                try await ditto.store.execute(query: query)
            } catch {
                print(
                    "TaskListScreenVM.\(#function) - ERROR evicting tasks: \(error.localizedDescription)"
                )
            }
        }
    }

    private nonisolated func populateTasksCollection() {
        Task {
            let initialTasks: [TaskModel] = [
                TaskModel(
                    _id: "50191411-4C46-4940-8B72-5F8017A04FA7",
                    title: "Buy groceries"),
                TaskModel(
                    _id: "6DA283DA-8CFE-4526-A6FA-D385089364E5",
                    title: "Clean the kitchen"),
                TaskModel(
                    _id: "5303DDF8-0E72-4FEB-9E82-4B007E5797F0",
                    title: "Schedule dentist appointment"),
                TaskModel(
                    _id: "38411F1B-6B49-4346-90C3-0B16CE97E174",
                    title: "Pay bills")
            ]

            for task in initialTasks {
                do {
                    try await ditto.store.execute(
                        query: "INSERT INTO tasks INITIAL DOCUMENTS (:task)",
                        arguments: [
                            "task":
                                [
                                    "_id": task._id,
                                    "title": task.title,
                                    "done": task.done,
                                    "deleted": task.deleted
                                ]
                        ]
                    )
                } catch {
                    print(
                        "TaskListScreenVM.\(#function) - ERROR creating initial task: \(error.localizedDescription)"
                    )
                }
            }
        }
    }

    func onEdit(task: TaskModel) {
        taskToEdit = task
        isPresentingEditScreen = true
    }

    func onNewTask() {
        taskToEdit = nil
        isPresentingEditScreen = true
    }

    func onBulkAdd() {
        isPresentingBulkAddScreen = true
    }
}
